File: ch4/kruskal_prim.ml
module UnionFind = struct
  type t =
    { parent: int array;
      rank: int array;
      size: int array;
      num_sets: int ref;
    }
  
  let create n =
    { parent  = Array.init n (fun x -> x);
      rank    = Array.make n 0;
      size    = Array.make n 1;
      num_sets = ref n;
    }
  
  let find { parent } u =
    let rec aux u =
      if parent.(u) = u then u
      else begin
        parent.(u) <- aux parent.(u);
        parent.(u)
      end
    in aux u
  
  let is_same_set dsu u v = find dsu u = find dsu v

  let union dsu u v =
    let link x p =
      dsu.parent.(x) <- p;
      dsu.size.(p) <- dsu.size.(p) + dsu.size.(x);
      if dsu.rank.(x) = dsu.rank.(p) then
        dsu.rank.(p) <- dsu.rank.(p) + 1
    in
    let pu, pv = find dsu u, find dsu v in
    if pu = pv then ()
    else begin
      decr dsu.num_sets;
      if dsu.rank.(pu) > dsu.rank.(pv) then link pv pu
      else link pu pv
    end
  
  let num_disjoint_sets { num_sets } = !num_sets

  let size_of_set dsu u = dsu.size.(find dsu u)
end

(** [kruskal n edges] returns the total weight of a MST of the input graph [(0..n-1, edges)] *)
let kruskal n edges =
  let dsu = UnionFind.create n in
  let cost = ref 0 in
  edges
    |> List.sort (fun (_, _, w1) (_, _, w2) -> compare w1 w2)
    |> List.iter (fun (u, v, w) ->
        if not (UnionFind.is_same_set dsu u v) then begin
          UnionFind.union dsu u v;
          cost := !cost + w
        end
      );
  !cost


module PQ = Set.Make (struct
  type t = int * int
  let compare = compare
end)

module IntS = Set.Make (struct
  type t = int
  let compare = compare
end)

(** [prim n edges] returns the total weight of a MST of the input graph [(0..n-1, edges)] *)
let prim n edges =
  let adj = Array.make n [] in
  edges |> List.iter (fun (u, v, w) ->
    adj.(u) <- (w, v) :: adj.(u);
    adj.(v) <- (w, u) :: adj.(v)
  );
  let rec loop taken candidates cost =
    match PQ.min_elt_opt candidates with
    | None -> cost
    | Some ((w, u) as e) ->
      let candidates = PQ.remove e candidates in
      if IntS.mem u taken then loop taken candidates cost
      else loop (IntS.add u taken)
                (candidates |> PQ.add_seq (List.to_seq adj.(u)))
                (cost + w)
  in loop (IntS.singleton 0) (PQ.of_list adj.(0)) 0

open Scanf
open Printf

let sc = object method nf = bscanf Scanning.stdin end

let () =
  let (n, m) = sc#nf "%d %d\n" (fun x y -> (x, y)) in
  let edges = List.init m (fun _ -> sc#nf "%d %d %d\n" (fun u v w -> (u, v, w))) in
  kruskal n edges |> printf "MST cost = %d (Kruskal's)\n";
  prim    n edges |> printf "MST cost = %d (Prim's)\n"

File: ch4/mst/kruskal_prim.py
from heapq import heappush, heappop

# Union-Find Disjoint Sets Library written in OOP manner
# using both path compression and union by rank heuristics
class UnionFind:                                # OOP style
    def __init__(self, N):
        self.p = [i for i in range(N)]
        self.rank = [0 for i in range(N)]
        self.setSize = [1 for i in range(N)]
        self.numSets = N

    def findSet(self, i):
        if (self.p[i] == i):
            return i
        else:
            self.p[i] = self.findSet(self.p[i])
            return self.p[i]

    def isSameSet(self, i, j):
        return self.findSet(i) == self.findSet(j)

    def unionSet(self, i, j):
        if (not self.isSameSet(i, j)):
            self.numSets -= 1
            x = self.findSet(i)
            y = self.findSet(j)
        # rank is used to keep the tree short
        if (self.rank[x] > self.rank[y]):
            self.p[y] = x
            self.setSize[x] += self.setSize[y]
        else:
            self.p[x] = y
            self.setSize[y] += self.setSize[x]
            if (self.rank[x] == self.rank[y]):
                self.rank[y] += 1

    def numDisjointSets(self):
        return self.numSets

    def sizeOfSet(self, i):
        return self.setSize[self.findSet(i)]

def process(u, taken, AL, pq):                      # set u as taken and enqueue neighbors of u
    taken[u] = 1
    for v, w in AL[u]:
        if (not taken[v]):
            heappush(pq, (w, v))                    # sort by (inc) weight
                                                    # then by (inc) id


def main():
    # Graph in Figure 4.10 left, format: list of weighted edges
    # This example shows another form of reading graph input
    # 5 7
    # 0 1 4
    # 0 2 4
    # 0 3 6
    # 0 4 6
    # 1 2 2
    # 2 3 8
    # 3 4 9

    f = open("kruskal_prim_in.txt", "r")

    V, E = map(int, f.readline().split(" "))
    # Kruskal's algorithm merged with Prim's algorithm
    AL = [[] for i in range(V)]                     # the graph stored in AL
    EL = []
    for i in range(E):
        u, v, w = map(int, f.readline().split(" ")) # read as (u, v, w)
        EL.append((w, u, v))                        # reorder as (w, u, v)
        AL[u].append((v, w))
        AL[v].append((u, w))
    EL.sort()                                       # sort by w, O(E log E)

    mst_cost = 0
    num_taken = 0
    UF = UnionFind(V)                               # all V are disjoint sets

    for i in range(E):                              # for each edge, O(E)
        if num_taken == V-1:
            break
        w, u, v = EL[i]
        if (not UF.isSameSet(u, v)):                # check
            num_taken += 1                          # 1 more edge is taken
            mst_cost += w                           # add w of this edge
            UF.unionSet(u, v)                       # link them
            # note: the runtime cost of UFDS is very light

    # note: the number of disjoint sets must eventually be 1 for a valid MST
    print("MST cost = {} (Kruskal's)".format(mst_cost))

    taken = [0 for i in range(V)]                   # to avoid cycle, no vertex is taken
    pq = []                                         # to select shorter edges
    process(0, taken, AL, pq)                       # take+process vertex 0
    mst_cost = 0                                    # no edge has been taken
    num_taken = 0
    while len(pq) > 0 and num_taken < V-1:          # until we take V-1 edges
        w, u = heappop(pq)
        if not taken[u]:                            # we have not taken u yet
            num_taken += 1                          # 1 more edge is taken
            mst_cost += w                           # add w of this edge
            process(u, taken, AL, pq)               # take+process vertex u
                                                    # each edge is in pq once

    print("MST cost = {} (Prim's)".format(mst_cost))

main()
